### PR TITLE
implementation of aggregated ReplicaDivisionPreference strategy

### PR DESCRIPTION
**Pull request by qianjun1993**
Signed-off-by: junqian <junqian@tencent.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Part of #344

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
None 

**How to work with it**:

Create deployment like this
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
      ···
```

and policy like this
```
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: example-policy
spec:
  resourceSelectors:
    - apiVersion: apps/v1
      kind: Deployment
      name: nginx
  placement:
    clusterAffinity:
      clusterNames:
        - member1
        - member2
    replicaScheduling:
      replicaSchedulingType: Divided
      replicaDivisionPreference: Aggregated
```

Then we will get a binding with the following clusters, result may be different according to the real resource of clusters
```
spec:
  clusters:
    - name: member1
       replicas : 3
    - name: member2
```

### CODE REVIEW

File: pkg/scheduler/core/generic_scheduler.go [hunk 1]
```diff
@@ -276,3 +282,125 @@ func (g *genericScheduler) divideReplicasByStaticWeight(clusters []*clusterv1alp
...
+	for i, cluster := range clusters {
+		maxReplicas := g.calClusterAvailableReplicas(cluster, replicaResourceRequirements)
+		availableTargetClusters[i] = workv1alpha1.TargetCluster{Name: cluster.Name, Replicas: maxReplicas}
+	}
+	sort.Sort(TargetClustersList(availableTargetClusters))
+	return availableTargetClusters
+}
+
+func (g *genericScheduler) calClusterAvailableReplicas(cluster *clusterv1alpha1.Cluster, resourcePerReplicas corev1.ResourceList) int32 {
+	ReplicasResult := int64(0)
```
Comment (mrlihanbo) [inline pkg/scheduler/core/generic_scheduler.go:330]: Why export the 'ReplicasResult '?

File: pkg/scheduler/core/generic_scheduler.go [hunk 2]
```diff
@@ -276,3 +282,125 @@ func (g *genericScheduler) divideReplicasByStaticWeight(clusters []*clusterv1alp
...
+	calFlag := false
```
Comment (mrlihanbo) [inline pkg/scheduler/core/generic_scheduler.go:331]: The name of this variable is not clear. Maybe we can name it like 'needUpdate'.     
@RainbowMango How do you think about it?

File: pkg/scheduler/core/generic_scheduler.go (hunk 1 above)
Comment (qianjun1993) [reply pkg/scheduler/core/generic_scheduler.go:330]: We need to know how many replicas  a cluster can schedule  according the resource.

File: pkg/scheduler/core/generic_scheduler.go (hunk 2 above)
Comment (qianjun1993) [reply pkg/scheduler/core/generic_scheduler.go:331]: You mean calFlag?

File: pkg/scheduler/core/generic_scheduler.go [hunk 3]
```diff
@@ -276,3 +282,125 @@ func (g *genericScheduler) divideReplicasByStaticWeight(clusters []*clusterv1alp
...
+			if !calFlag {
+				ReplicasResult = maxReplicas
+				calFlag = true
+			} else if ReplicasResult > maxReplicas {
+				ReplicasResult = maxReplicas
+			}
+		}
+	}
+	return int32(ReplicasResult)
+}
```
Comment (RainbowMango) [inline pkg/scheduler/core/generic_scheduler.go:362]:
```suggestion
// calClusterAvailableReplicas calculates how many replicas can be applied to the target cluster.
func (g *genericScheduler) calClusterAvailableReplicas(cluster *clusterv1alpha1.Cluster, resourcePerReplicas corev1.ResourceList) int32 {
	var maximumReplicas int64 = math.MaxInt32
	resourceSummary := cluster.Status.ResourceSummary

	for key, value := range resourcePerReplicas {
		requestedQuantity := value.Value()
		if requestedQuantity <= 0 {
			continue
		}

		// calculates available resource quantity
		// available = allocatable - allocated - allocating
		allocatable, ok := resourceSummary.Allocatable[key]
		if !ok {
			return 0
		}
		allocated, ok := resourceSummary.Allocated[key]
		if ok {
			allocatable.Sub(allocated)
		}
		allocating, ok := resourceSummary.Allocating[key]
		if ok {
			allocatable.Sub(allocating)
		}
		availableQuantity := allocatable.Value()
		// short path: no more resource left.
		if availableQuantity <= 0 {
			return 0
		}

		maximumReplicasForResource := availableQuantity / requestedQuantity
		if maximumReplicasForResource < maximumReplicas {
			maximumReplicas = maximumReplicasForResource
		}
	}

	return int32(maximumReplicas)
}
```

Comment (RainbowMango) [reply pkg/scheduler/core/generic_scheduler.go:362]: How about this?

**Review by RainbowMango — CHANGES_REQUESTED** [review]

File: pkg/scheduler/core/generic_scheduler.go (hunk 3 above)
Comment (qianjun1993) [reply pkg/scheduler/core/generic_scheduler.go:362]: OK

File: pkg/scheduler/core/generic_scheduler.go [hunk 4]
```diff
@@ -276,3 +283,130 @@ func (g *genericScheduler) divideReplicasByStaticWeight(clusters []*clusterv1alp
...
+	return availableTargetClusters
+}
+
+// calClusterAvailableReplicas calculates how many replicas can be applied to the target cluster.
+func (g *genericScheduler) calClusterAvailableReplicas(cluster *clusterv1alpha1.Cluster, resourcePerReplicas corev1.ResourceList) int32 {
+	var maximumReplicas int64 = math.MaxInt32
+	resourceSummary := cluster.Status.ResourceSummary
+
+	for key, value := range resourcePerReplicas {
+		requestedQuantity := value.Value()
```
Comment (RainbowMango) [inline pkg/scheduler/core/generic_scheduler.go:336]: @pigletfly 
Should we use `Value()` or `MilliValue()` here?

Comment (RainbowMango) [reply pkg/scheduler/core/generic_scheduler.go:336]: @qianjun1993 
By the way, I think we need a unit test here.

Comment (qianjun1993) [reply pkg/scheduler/core/generic_scheduler.go:336]: It should be MilliValue when the resource is 'CPU'

**Comment by RainbowMango** [comment]
Record a test case:

I have 3 clusters(member1, member2, member3), and each have 4 core
```yaml
  resourceSummary:
    allocatable:
      cpu: "4"
      ephemeral-storage: 123722284Ki
      hugepages-1Gi: "0"
      hugepages-2Mi: "0"
      memory: 16265856Ki
      pods: "110"
    allocated:
      cpu: 850m
      memory: 190Mi
    allocating:
      cpu: "0"
      memory: "0"
```

Deploy a deployment that requests `1` CPU for each replica:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  replicas: 6
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx
        name: nginx
        resources:
          requests:
            cpu: "1"
```

Deploy a policy that expect aggregate the replicas:
```yaml
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: nginx-propagation
spec:
  resourceSelectors:
    - apiVersion: apps/v1
      kind: Deployment
      name: nginx
  placement:
    clusterAffinity:
      clusterNames:
        - member1
        - member2
        - member3
    replicaScheduling:
      replicaSchedulingType: Divided
      replicaDivisionPreference: Aggregated
```

The schedule result:
```yaml
apiVersion: work.karmada.io/v1alpha1
kind: ResourceBinding
metadata:
  annotations:
    policy.karmada.io/applied-placement: '{"clusterAffinity":{"clusterNames":["member1","member2","member3"]},"replicaScheduling":{"replicaSchedulingType":"Divided","replicaDivisionPreference":"Aggregated"}}'
  creationTimestamp: "2021-07-10T02:47:01Z"
  generation: 10
  labels:
    propagationpolicy.karmada.io/name: nginx-propagation
    propagationpolicy.karmada.io/namespace: default
  name: nginx-deployment
  namespace: default
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: Deployment
    name: nginx
    uid: acfaa4e7-7a4c-423a-89d9-df33aac6781c
  resourceVersion: "2863"
  selfLink: /apis/work.karmada.io/v1alpha1/namespaces/default/resourcebindings/nginx-deployment
  uid: f9994ed1-290b-4738-826e-e30e8da26b64
spec:
  clusters:
  - name: member1
    replicas: 3
  - name: member2
    replicas: 3
  - name: member3
  resource:
    apiVersion: apps/v1
    kind: Deployment
    name: nginx
    namespace: default
    replicas: 6
    resourcePerReplicas:
      cpu: "1"
      memory: "0"
    resourceVersion: "2861"
```

**Pull request merged**